apps/manager_app/auth_card.c

```diff
@@ -465,118 +358,140 @@ static manager_error_code_t handle_sign_challenge(
-      if (false == result) {
-        resp->which_response = MANAGER_AUTH_CARD_RESPONSE_FLOW_COMPLETE_TAG;
-        resp->flow_complete.dummy_field = 0;
+      if (false == verified) {
+        result.auth_card.which_response =
+            MANAGER_AUTH_CARD_RESPONSE_FLOW_COMPLETE_TAG;
+        result.auth_card.flow_complete.dummy_field = 0;
+        manager_send_result(&result);
+
         delay_scr_init(ui_text_card_authentication_failed, DELAY_TIME);
-        return MANAGER_TASK_SUCCESS;
+        return true;
       } else {
-        return MANAGER_TASK_INVALID_STATE;
+        manager_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                           ERROR_DATA_FLOW_INVALID_REQUEST);
+        return false;
       }
       break;
 
     case MANAGER_AUTH_CARD_STATUS_CHALLENGE_SIGNED:
-      if (false == result) {
-        resp->which_response = MANAGER_AUTH_CARD_RESPONSE_FLOW_COMPLETE_TAG;
-        resp->flow_complete.dummy_field = 0;
+      if (false == verified) {
+        result.auth_card.which_response =
+            MANAGER_AUTH_CARD_RESPONSE_FLOW_COMPLETE_TAG;
+        result.auth_card.flow_complete.dummy_field = 0;
+        manager_send_result(&result);
+
         delay_scr_init(ui_text_card_authentication_failed, DELAY_TIME);
-        return MANAGER_TASK_SUCCESS;
+        return true;
       } else {
         if (true == auth_card_data->ctx.pair_card_required) {
           uint8_t card_number =
               decode_card_number(auth_card_data->ctx.acceptable_cards);
-          bool result = pair_card_operation(card_number,
-                                            auth_card_data->ctx.heading,
-                                            auth_card_data->ctx.message);
-
-          if (true != result) {
-            delay_scr_init("Pairing operation failed", DELAY_TIME);
-            return MANAGER_TASK_FAILED;
+          bool pairing_done = pair_card_operation(card_number,
+                                                  auth_card_data->ctx.heading,
+                                                  auth_card_data->ctx.message);
+
+          if (true != pairing_done) {
+            manager_send_error(ERROR_COMMON_ERROR_CARD_ERROR_TAG,
+                               ERROR_CARD_ERROR_UNKNOWN);
+            delay_scr_init(ui_text_card_authentication_failed, DELAY_TIME);
+            return false;
           }
         }
 
-        core_status_set_flow_status(MANAGER_AUTH_CARD_STATUS_PAIRING_DONE);
-        resp->which_response = MANAGER_AUTH_CARD_RESPONSE_FLOW_COMPLETE_TAG;
-        resp->flow_complete.dummy_field = 0;
+        UPDATE_FLOW_STATUS(auth_card_data,
+                           MANAGER_AUTH_CARD_STATUS_PAIRING_DONE);
+        result.auth_card.which_response =
+            MANAGER_AUTH_CARD_RESPONSE_FLOW_COMPLETE_TAG;
+        result.auth_card.flow_complete.dummy_field = 0;
+        manager_send_result(&result);
+
+        if (MANAGER_ONBOARDING_STEP_COMPLETE != onboarding_get_last_step()) {
```

Comment:
No need to check for this, the write API will protect within itself

Reply:
right will remove

Reply:
removed